### PR TITLE
Auto import needed packages

### PR DESCRIPTION
**Pull request by eugene-auduchinok**
It can be disabled from Settings | Editor | General | Auto import, however it needs Java auto import enabled.

### CODE REVIEW

**Comment by eugene-auduchinok** [comment]
Needs adding tests, though.

File: src/com/goide/codeInsight/imports/GoCodeInsightSettings.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import com.intellij.openapi.components.*;
+import com.intellij.util.xmlb.XmlSerializerUtil;
+import org.jetbrains.annotations.Nullable;
+
+@State(
+  name = "GoCodeInsightSettings",
+  storages = @Storage(file = StoragePathMacros.APP_CONFIG + "/editor.codeinsight.xml")
+)
+public class GoCodeInsightSettings implements PersistentStateComponent<GoCodeInsightSettings> {
+  public boolean ADD_UNAMBIGUOUS_IMPORTS_ON_THE_FLY = true;
```
Comment (zolotov) [inline src/com/goide/codeInsight/imports/GoCodeInsightSettings.java]: incapsulate

File: src/com/goide/psi/impl/imports/GoReferenceImporter.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    final PsiReference reference = file.findReferenceAt(offset);
+    if (reference instanceof GoReference) {
+      GoImportPackageQuickFix fix = new GoImportPackageQuickFix(reference);
+      if (fix.isAvailable(file.getProject(), editor, file)) {
+        WriteCommandAction.runWriteCommandAction(file.getProject(), new Runnable() {
+          @Override
+          public void run() {
+            new GoImportPackageQuickFix(reference).applyFix();
```
Comment (zolotov) [inline src/com/goide/psi/impl/imports/GoReferenceImporter.java]: why do you need second quickFix?
are you sure that fix is still applicable when write actions is started?

File: src/com/goide/psi/impl/imports/GoReferenceImporter.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    int endOffset = document.getLineEndOffset(lineNumber);
+
+    List<PsiElement> elements = CollectHighlightsUtil.getElementsInRange(file, startOffset, endOffset);
+    for (PsiElement element : elements) {
+      if (element instanceof GoElement) {
+        for (PsiReference reference : element.getReferences()) {
+          if (reference instanceof GoReference) {
+            GoImportPackageQuickFix fix = new GoImportPackageQuickFix(reference);
+            if (fix.isAvailable(file.getProject(), editor, file)) {
+              new GoImportPackageQuickFix(reference).applyFix();
```
Comment (zolotov) [inline src/com/goide/psi/impl/imports/GoReferenceImporter.java]: Are you sure that you don't need write lock here?

File: src/com/goide/codeInsight/imports/GoAutoImportOptions.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  @Nullable
+  @Override
+  public JComponent createComponent() {
+    return myPanel;
+  }
+
+  @Override
+  public boolean isModified() {
+    GoCodeInsightSettings settings = GoCodeInsightSettings.getInstance();
```
Comment (zolotov) [inline src/com/goide/codeInsight/imports/GoAutoImportOptions.java]: I believe you can use service auto-injection. Just add GoCodeInsightSettings as a constructor parameter

**Comment by dlsniper** [comment]
@zolotov  / @eugene-auduchinok as a quick question, will this work for the small IDEs / Android Studio as well?

**Comment by zolotov** [comment]
Didn't checked, but there is no any IDEA specific things so it should

**Comment by dlsniper** [comment]
ok, I'll try to investigate later. Thanks

**Pull request closed without merging**